### PR TITLE
Produce full pages from local partitioned exchange

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 1]
```diff
@@ -62,37 +63,46 @@ public PartitioningExchanger(
...
+        flush(false);
+    }
+
+    @Override
+    public void flush()
+    {
+        flush(true);
+    }
+
+    private void appendRow(PageBuilder pageBuilder, Page page, int position)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: Are there any benchmarks for exchange performance? Is appending row-by-row any slower than using copyPositions, in exchange for having full pages of output that makes subsequent execution faster?

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 2]
```diff
@@ -62,37 +63,46 @@ public PartitioningExchanger(
...
+    public void flush(boolean force)
+    {
+        for (int partition = 0; partition < pageBuilders.size(); partition++) {
+            PageBuilder partitionPageBuilder = pageBuilders.get(partition);
+            if (!partitionPageBuilder.isEmpty() && (force || partitionPageBuilder.isFull())) {
+                Page pageSplit = partitionPageBuilder.build();
+                partitionPageBuilder.reset();
                 memoryTracker.accept(pageSplit.getRetainedSizeInBytes());
                 buffers.get(partition).accept(new PageReference(pageSplit, 1, () -> memoryTracker.accept(-pageSplit.getRetainedSizeInBytes())));
             }
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: You need to keep track of the memory that is in the non-flushed `PageBuilders`, too.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/Exchanger.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.spi.Page;
+
+interface Exchanger
+{
+    void accept(Page page);
+
+    default void flush() {}
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/Exchanger.java]: Please add javadoc for these methods, `flush()` in particular.

File: presto-main/src/test/java/com/facebook/presto/operator/exchange/TestLocalExchange.java [hunk 4]
```diff
@@ -235,13 +238,17 @@ public void testPartition()
         LocalExchangeSource sourceB = exchange.getSource(1);
         assertSource(sourceB, 0);
 
-        sink.addPage(createPage(0));
```
Comment (cawallin) [inline presto-main/src/test/java/com/facebook/presto/operator/exchange/TestLocalExchange.java]: Can you add a test where you add only a partial page, and where you have values that are in different partitions?

**Comment by cawallin** [comment]
Both Travis failures are memory related, not sure if they are intermittent or related to this patch.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java (hunk 1 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: It mustn't affect the performance. Both approached do data copying from random position in a source block.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java (hunk 2 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: Good point. I'm going to track the memory on input. Partitioned exchange doesn't filter out any results. So the size of buffered pages should be equal to the size of the input page.

File: presto-main/src/test/java/com/facebook/presto/operator/exchange/TestLocalExchange.java (hunk 4 above)
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/operator/exchange/TestLocalExchange.java]: Discussed offline

File: presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java [hunk 5]
```diff
@@ -78,9 +78,7 @@ public void testBlocking()
         }
 
         // driver should be blocked waiting for memory
-        for (Driver driver : drivers) {
-            assertFalse(driver.isFinished());
-        }
+        assertTrue(isWaitingForMemory(drivers));
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java]: As local exchange now buffers data to produce full pages drivers with TableScan -> LocalExchangeSink is not always instantly blocked by the drivers above.

**Comment by nezihyigitbasi** [comment]
What's the motivation for this change, and do you have any numbers showing that this change has a performance advantage/disadvantage?

**Comment by arhimondr** [comment]
> What's the motivation for this change

Here is the another related PR: https://github.com/prestodb/presto/pull/8691

If the input pages are tiny - this exchange makes them even smaller. It results in additional synchronization overhead.

Here is an example of query we had problems with: 

```
SELECT l.extendedprice * (1 - l.discount) AS revenue 
FROM lineitem AS l, orders AS o 
WHERE l.orderkey = o.orderkey 
AND o.orderdate >= DATE '1993-10-01' 
AND o.orderdate < DATE '1993-10-01' + INTERVAL '3' MONTH AND l.shipmode='NONE'
```

Build side filter is very selective. Cost based optimizer chooses REPLICATED join type for this query as less expensive. Because of the bug fixed in #8691 local partitioning exchange was getting tiny pages as an input. So, after local partitioning, HashBuilder operator was getting even smaller pages (empty pages, few line pages). HashBuilderOperator reserves memory for every single page, so the overhead was quite considerable.

After introducing this fix we managed to cut execiton time of that query on `1TB` `TPC-DS` schema from `1m10s` to `40s`. 

After applying the fix from #8691 the execution time of this query has been reduced to `10-11s`.

> Do you have any numbers showing that this change has a performance advantage/disadvantage?

In theory there shouldn't be any slowdowns after introducing this patch. The only advantage of using `Block#copyPositions` over iterating over the rows manually is when a Block is `DictionaryBlock` or `RunLenghtBlock`. But those types of block will be converted to a regular block at the Remote re-partitioning step.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 6]
```diff
@@ -62,40 +63,61 @@ public PartitioningExchanger(
...
+            type.appendTo(page.getBlock(channel), position, pageBuilder.getBlockBuilder(channel));
+        }
+    }
 
-                Page pageSplit = new Page(positions.size(), outputBlocks);
+    private void flush(boolean force)
+    {
+        for (int partition = 0; partition < pageBuilders.size(); partition++) {
+            PageBuilder partitionPageBuilder = pageBuilders.get(partition);
+            if (!partitionPageBuilder.isEmpty() && (force || partitionPageBuilder.isFull())) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: I would split this if into two.
If builder empty then continue. If force or full then produce/consume.
Up to you though.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 7]
```diff
@@ -62,40 +63,61 @@ public PartitioningExchanger(
...
-        // build a page for each partition
-        Block[] sourceBlocks = page.getBlocks();
-        Block[] outputBlocks = new Block[sourceBlocks.length];
-        for (int partition = 0; partition < buffers.size(); partition++) {
-            List<Integer> positions = partitionAssignments[partition];
-            if (!positions.isEmpty()) {
-                for (int i = 0; i < sourceBlocks.length; i++) {
-                    outputBlocks[i] = sourceBlocks[i].copyPositions(positions);
-                }
+    private void appendRow(PageBuilder pageBuilder, Page page, int position)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: I would put this method one up or down, to keep flush methods together

**Comment by losipiuk** [comment]
> But those types of block will be converted to a regular block at the Remote re-partitioning step.

**Comment by sopel39** [comment]
I don't think we should rush with this change without fully checking it's impact as we know it influences both memory and performance. Maybe https://github.com/prestodb/presto/pull/8691 (with comments applied) would be enough for now.

> The only advantage of using Block#copyPositions over iterating over the rows manually is when a Block is DictionaryBlock or RunLenghtBlock. But those types of block will be converted to a regular block at the Remote re-partitioning step.

That's only true because currently `PartitionedOutputBuffer` uses `PageBuilder`. However, one could easily implement support for `DictionaryBlock` in `PartitionedOutputBuffer`. Therefore dictionaries would be sent over network and `copyPositions` could be implemented efficiently.

**Comment by sopel39** [comment]
> I have a PR I'm working on that does not copy at all and instead uses DictionaryBlock to mask the input page for each output partition.

Aren't those two separate/independent issues:
* using dictionaries for masking
* merging tiny pages
?

We probably want to merge extremely small pages, so that merging mechanism doesn't always kick in. I understand that you want to hold this PR until your `DictionaryBlock` masking lands in?

**Comment by arhimondr** [comment]
> As discussed with @sopel39, your PR will reintroduce the higher memory usage for join build side as BlockBuilder allocates 1.25 more memory than requested and in case of variable length types blocks in page might not be fully utilized causing even more memory wasting.

HashBuilderOperator does the compaction as part of the index building process.

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 8]
```diff
@@ -62,40 +63,61 @@ public PartitioningExchanger(
...
+    {
+        pageBuilder.declarePosition();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: > com.facebook.presto.spi.block.Block#writeBytesTo

I think i don't understand this method. `writeBytesTo` allows you to copy chunk of value in a position. Not the entire block. Am i wrong?

File: presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java [hunk 9]
```diff
@@ -17,28 +17,28 @@
...
 import static java.util.Objects.requireNonNull;
 
 class PartitioningExchanger
-        implements Consumer<Page>
+        implements Exchanger
 {
     private final List<Consumer<PageReference>> buffers;
     private final LongConsumer memoryTracker;
     private final LocalPartitionGenerator partitionGenerator;
-    private final IntList[] partitionAssignments;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java]: I don't use any collection now. I assign partitions on the go.

**Comment by sopel39** [comment]
>HashBuilderOperator does the compaction as part of the index building process.

HashBuilderOperator does the compaction only when the machine is running out of memory. Otherwise no compaction occurs. Therefore this limits cluster concurrency, because first join queries (that don't run out of memory) won't compact pages and occupy extra unused memory.

**Comment by arhimondr** [comment]
I just checked the #8691 without this patch. It seems that that patch alone solves the problem. Considering that @dain is working in completely opposite direction we can close this PR for now.

**Comment by sopel39** [comment]
> I have a PR I'm working on that does not copy at all and instead uses DictionaryBlock to mask the input page for each output partition.

@dain please ping me when the PR is ready. Do you have some WIP?

**Pull request closed without merging**